daemon/k8s_watcher.go

```diff
@@ -930,13 +995,39 @@ func (d *Daemon) addK8sEndpointV1(ep *v1.Endpoints) {
 	}
 
 	svc, ok := d.loadBalancer.K8sServices[svcns]
+
 	if ok && svc.IsExternal() {
-		translator := k8s.NewK8sTranslator(svcns, *newSvcEP, false, svc.Labels, bpfIPCache.IPCache)
-		err := d.policy.TranslateRules(translator)
-		if err != nil {
-			log.Errorf("Unable to repopulate egress policies from ToService rules: %v", err)
-		} else {
-			d.TriggerPolicyUpdates(true, "Kubernetes service endpoint added")
+		serviceImportMeta, cacheOK := endpointMetadataCache.get(ep)
+
+		// If this is the first time adding this Endpoint, or there was an error
+		// adding it last time, then try to add translate it and its
+		// corresponding external service for any toServices rules which
+		// select said service.
+		if !cacheOK || (cacheOK && serviceImportMeta.ruleTranslationError != nil) {
+			translator := k8s.NewK8sTranslator(svcns, *newSvcEP, false, svc.Labels, bpfIPCache.IPCache)
+			err := d.policy.TranslateRules(translator)
+			endpointMetadataCache.upsert(ep, err)
```

Comment:
I think this should really be a controller but I understand that this is an incremental update and the purpose here is to avoid unnecessary regeneration calls so I think this is a step in the right direction.